pkg/controller/worker/state_actuator.go

```diff
@@ -0,0 +1,208 @@
+
+		machineDeploymentState.Replicas = deployment.Spec.Replicas
+
+		machineSet, ok := machineSets[deployment.Name]
+		if !ok {
+			continue
+		}
+		addMachineSetToMachineDeploymentState(machineSet, machineDeploymentState)
+
+		currentMachines := append(machines[machineSet.Name], machines[deployment.Name]...)
+		if len(currentMachines) <= 0 {
+			continue
+		}
+
+		for index := range currentMachines {
+			addMachineToMachineDeploymentState(&currentMachines[index], machineDeploymentState)
+		}
+
+		workerState[deployment.Name] = machineDeploymentState
+	}
+
+	return json.Marshal(workerState)
+}
+
+// getExistingMachineSetsMap returns a map of existing MachineSets as values and their owners as keys
+func (a *genericStateActuator) getExistingMachineSetsMap(ctx context.Context, worker *extensionsv1alpha1.Worker) (map[string]*machinev1alpha1.MachineSet, error) {
+	existingMachineSets := &machinev1alpha1.MachineSetList{}
+	if err := a.client.List(ctx, existingMachineSets, client.InNamespace(worker.Namespace)); err != nil {
+		return nil, err
+	}
+
+	machineSets := make(map[string]*machinev1alpha1.MachineSet)
+	for index, machineSet := range existingMachineSets.Items {
+		if len(machineSet.OwnerReferences) > 0 {
+			for _, referant := range machineSet.OwnerReferences {
+				if referant.Kind == machineDeploymentKind {
+					machineSets[referant.Name] = &existingMachineSets.Items[index]
+				}
+			}
+		} else if len(machineSet.Labels) > 0 {
+			if machineDeploymentName, ok := machineSet.Labels[machineDeploymentLabel]; ok {
+				machineSets[machineDeploymentName] = &existingMachineSets.Items[index]
+			}
+		}
+	}
+	return machineSets, nil
+}
+
+// getExistingMachinesMap returns a map of the existing Machines as values and the name of their owner
+// no matter of being machineSet or MachineDeployment. If a Machine has a ownerRefernce the key(owner)
+// will be the MachineSet if not the key will be the name of the MachineDeployment which is stored as
+// a lable. We assume that there is no MachineDeployment and MachineSet with the same names.
+func (a *genericStateActuator) getExistingMachinesMap(ctx context.Context, worker *extensionsv1alpha1.Worker) (map[string][]machinev1alpha1.Machine, error) {
+	existingMachines := &machinev1alpha1.MachineList{}
+	if err := a.client.List(ctx, existingMachines, client.InNamespace(worker.Namespace)); err != nil {
+		return nil, err
+	}
+
+	machines := make(map[string][]machinev1alpha1.Machine)
+	for index, machine := range existingMachines.Items {
+		if len(machine.OwnerReferences) > 0 {
+			for _, referant := range machine.OwnerReferences {
+				if referant.Kind == machineSetKind {
+					machines[referant.Name] = append(machines[referant.Name], existingMachines.Items[index])
+				}
+			}
+		} else if len(machine.Labels) > 0 {
+			if machineDeploymentName, ok := machine.Labels[machineDeploymentLabel]; ok {
+				machines[machineDeploymentName] = append(machines[machineDeploymentName], existingMachines.Items[index])
```

Comment:
@rfranzke 
This is a ugly avoidance of not getting in account an orphaned machine. If a Machine has `OwnerReference` this will be the `MachineSet`. but If there is no owner then we can use the lable which represents the `MachineDeployment`